core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -375,1083 +380,674 @@ public boolean isPreloaded() {
+               if (trace) {
+                  log.tracef("Publishing keys for segments %s", segments);
+               }
+               NonBlockingStore<K, ?> nonBlockingStore = getStoreLocked(storeStatus ->
+                     storeStatus.characteristics.contains(Characteristic.BULK_READ) && predicate.test(storeStatus.config));
+               return nonBlockingStore == null ? Flowable.empty() : nonBlockingStore.publishKeys(segments, filter);
+            },
+            this::releaseReadLock);
    }
 
    @Override
-   public <K> Flowable<K> publishKeys(IntSet segments, Predicate<? super K> filter,
-         Predicate<? super StoreConfiguration> predicate) {
-      SegmentedAdvancedLoadWriteStore<K, ?> segmentedStore = getFirstSegmentedStore(predicate);
-
-      if (segmentedStore != null) {
-         // We have to acquire the read lock on the stores mutex to be sure that no concurrent stop or store removal
-         // is done while processing data
-         return Flowable
-               .using(publisherSemaphoreCallable,
-                     semaphore -> handler.blockingPublisher(segmentedStore.publishKeys(segments, filter)),
-                     Semaphore::release);
-      }
-
-      return publishKeys(PersistenceUtil.combinePredicate(segments, keyPartitioner, filter), predicate);
+   public <K, V> CompletionStage<MarshallableEntry<K, V>> loadFromAllStores(Object key, boolean localInvocation,
+         boolean includeStores) {
+      return loadFromAllStores(key, keyPartitioner.getSegment(key), localInvocation, includeStores);
    }
 
-   private <K, V> MarshallableEntry<K, V> loadFromAllStoresSync(Object key, boolean localInvocation, boolean includeStores, int traceId) {
-      acquireReadLock();
-      try {
-         checkStoreAvailability();
-         if (trace) {
-            log.tracef("Loading entry for key %s from stores with includeStores %s for id: %d",
-                  key, includeStores, traceId);
-         }
-         MarshallableEntry load = null;
-         for (CacheLoader l : loaders) {
-            if (allowLoad(l, localInvocation, includeStores)) {
-               load = l.loadEntry(key);
-               if (load != null)
-                  break;
-            }
-         }
-         if (trace) {
-            log.tracef("Entry was loaded: %s for key %s from stores for id: %d", load, key, traceId);
-         }
-         return load;
-      } finally {
-         releaseReadLock();
+   @Override
+   public <K, V> CompletionStage<MarshallableEntry<K, V>> loadFromAllStores(Object key, int segment,
+         boolean localInvocation, boolean includeStores) {
+      return Maybe.using(
+            this::acquireReadLock,
+            ignore -> {
+               checkStoreAvailability();
+               if (trace) {
+                  log.tracef("Loading entry for key %s with segment %d", key, segment);
+               }
+               return Flowable.fromIterable(stores.entrySet())
```

Comment:
`StoreStatus` contains a reference to the store. you could use `stores.values()`.
And, couple line below, just do `value.store().load(segment, key)`
Also, I'm wondering the `stores` map can be converted to a list or array.

Reply:
I had debated about changing it to a List at one point, but didn't mess with the idea too much. I can try to bake something up real quick. But tbh I would rather this not prevent integration, unless there is something else more important.

Reply:
I'm ok to be merged as is @wburns  :) I've finished my quick look around :+1:

Reply:
Thanks a ton, I am trying to finish up the conversion to ArrayList real quick if that is it.